Isolation: fixed mapping of user and group IDs.

alejandro-colomar opened the pull request:
Closes: <https://github.com/nginx/unit/issues/788>
Closes: <https://github.com/nginx/unit/issues/799>

alejandro-colomar commented:
@bsidhu009

You could try this fix.  It works for me.  :)

ac000 commented on src/nxt_clone.c:
```diff
@@ -387,8 +387,25 @@ nxt_clone_vldt_credential_gidmap(nxt_task_t *task,
+        m = map->map[i];
 
+        if (creds->base_gid >= (nxt_gid_t) m.container
+            && creds->base_gid < (nxt_gid_t) (m.container+m.size))
```
I know it was like this already, but the general style seems to be to have a space around numerical operators.

alejandro-colomar replied:
Now I've force pushed to change the patch subject, to clarify that it only _moves_ code.  No changes; not even whitespace.

ac000 replied:
Ah, OK.

alejandro-colomar commented on src/nxt_clone.c:
```diff
@@ -460,4 +456,17 @@ nxt_id_map_host2container(id_t host_id, const nxt_clone_map_entry_t *map)
 }
 
+
+static inline
+nxt_bool_t
+nxt_idmap_includes_id(id_t host_id, const nxt_clone_map_entry_t *map)
```
I just wrote from memory; I thought we used that style.   But I just checked, and it goes in the same line as the type.  It's attributes that go on a separate line.  I'll fix this.

Just for the curious:

```c
$ find src/nxt_* -type f | grep '\.[ch]$' | xargs grep -nA1 'static inline'
src/nxt_app_nncq.h:24:static inline nxt_app_nncq_atomic_t
src/nxt_app_nncq.h-25-nxt_app_nncq_head(nxt_app_nncq_t const volatile *q)
--
src/nxt_app_nncq.h:31:static inline nxt_app_nncq_atomic_t
src/nxt_app_nncq.h-32-nxt_app_nncq_tail(nxt_app_nncq_t const volatile *q)
--
src/nxt_app_nncq.h:38:static inline void
src/nxt_app_nncq.h-39-nxt_app_nncq_tail_cmp_inc(nxt_app_nncq_t volatile *q, nxt_app_nncq_atomic_t t)
--
src/nxt_app_nncq.h:45:static inline nxt_app_nncq_atomic_t
src/nxt_app_nncq.h-46-nxt_app_nncq_index(nxt_app_nncq_t const volatile *q, nxt_app_nncq_atomic_t i)
--
src/nxt_app_nncq.h:52:static inline nxt_app_nncq_atomic_t
src/nxt_app_nncq.h-53-nxt_app_nncq_map(nxt_app_nncq_t const volatile *q, nxt_app_nncq_atomic_t i)
--
src/nxt_app_nncq.h:59:static inline nxt_app_nncq_cycle_t
src/nxt_app_nncq.h-60-nxt_app_nncq_cycle(nxt_app_nncq_t const volatile *q, nxt_app_nncq_atomic_t i)
--
src/nxt_app_nncq.h:66:static inline nxt_app_nncq_cycle_t
src/nxt_app_nncq.h-67-nxt_app_nncq_next_cycle(nxt_app_nncq_t const volatile *q,
--
src/nxt_app_nncq.h:74:static inline nxt_app_nncq_atomic_t
src/nxt_app_nncq.h-75-nxt_app_nncq_new_entry(nxt_app_nncq_t const volatile *q,
--
src/nxt_app_nncq.h:83:static inline nxt_app_nncq_atomic_t
src/nxt_app_nncq.h-84-nxt_app_nncq_empty(nxt_app_nncq_t const volatile *q)
--
src/nxt_clang.h:11:#define nxt_inline     static inline __attribute__((always_inline))
src/nxt_clang.h-12-#define nxt_noinline   __attribute__((noinline))
--
src/nxt_nncq.h:24:static inline nxt_nncq_atomic_t
src/nxt_nncq.h-25-nxt_nncq_head(nxt_nncq_t const volatile *q)
--
src/nxt_nncq.h:31:static inline nxt_nncq_atomic_t
src/nxt_nncq.h-32-nxt_nncq_tail(nxt_nncq_t const volatile *q)
--
src/nxt_nncq.h:38:static inline void
src/nxt_nncq.h-39-nxt_nncq_tail_cmp_inc(nxt_nncq_t volatile *q, nxt_nncq_atomic_t t)
--
src/nxt_nncq.h:45:static inline nxt_nncq_atomic_t
src/nxt_nncq.h-46-nxt_nncq_index(nxt_nncq_t const volatile *q, nxt_nncq_atomic_t i)
--
src/nxt_nncq.h:52:static inline nxt_nncq_atomic_t
src/nxt_nncq.h-53-nxt_nncq_map(nxt_nncq_t const volatile *q, nxt_nncq_atomic_t i)
--
src/nxt_nncq.h:59:static inline nxt_nncq_cycle_t
src/nxt_nncq.h-60-nxt_nncq_cycle(nxt_nncq_t const volatile *q, nxt_nncq_atomic_t i)
--
src/nxt_nncq.h:66:static inline nxt_nncq_cycle_t
src/nxt_nncq.h-67-nxt_nncq_next_cycle(nxt_nncq_t const volatile *q, nxt_nncq_cycle_t i)
--
src/nxt_nncq.h:73:static inline nxt_nncq_atomic_t
src/nxt_nncq.h-74-nxt_nncq_new_entry(nxt_nncq_t const volatile *q, nxt_nncq_cycle_t cycle,
--
src/nxt_nncq.h:81:static inline nxt_nncq_atomic_t
src/nxt_nncq.h-82-nxt_nncq_empty(nxt_nncq_t const volatile *q)
--
src/nxt_nncq.h:88:static inline void
src/nxt_nncq.h-89-nxt_nncq_init(nxt_nncq_t volatile *q)
--
src/nxt_nncq.h:97:static inline void
src/nxt_nncq.h-98-nxt_nncq_enqueue(nxt_nncq_t volatile *q, nxt_nncq_atomic_t val)
--
src/nxt_nncq.h:131:static inline nxt_nncq_atomic_t
src/nxt_nncq.h-132-nxt_nncq_dequeue(nxt_nncq_t volatile *q)
--
src/nxt_nvbcq.h:25:static inline nxt_nvbcq_atomic_t
src/nxt_nvbcq.h-26-nxt_nvbcq_head(nxt_nvbcq_t const volatile *q)
--
src/nxt_nvbcq.h:32:static inline nxt_nvbcq_atomic_t
src/nxt_nvbcq.h-33-nxt_nvbcq_tail(nxt_nvbcq_t const volatile *q)
--
src/nxt_nvbcq.h:39:static inline void
src/nxt_nvbcq.h-40-nxt_nvbcq_tail_cmp_inc(nxt_nvbcq_t volatile *q, nxt_nvbcq_atomic_t t)
--
src/nxt_nvbcq.h:46:static inline nxt_nvbcq_atomic_t
src/nxt_nvbcq.h-47-nxt_nvbcq_index(nxt_nvbcq_t const volatile *q, nxt_nvbcq_atomic_t i)
--
src/nxt_nvbcq.h:53:static inline nxt_nvbcq_atomic_t
src/nxt_nvbcq.h-54-nxt_nvbcq_map(nxt_nvbcq_t const volatile *q, nxt_nvbcq_atomic_t i)
--
src/nxt_nvbcq.h:60:static inline nxt_nvbcq_atomic_t
src/nxt_nvbcq.h-61-nxt_nvbcq_empty(nxt_nvbcq_t const volatile *q)
--
src/nxt_nvbcq.h:67:static inline void
src/nxt_nvbcq.h-68-nxt_nvbcq_init(nxt_nvbcq_t volatile *q)
--
src/nxt_nvbcq.h:82:static inline void
src/nxt_nvbcq.h-83-nxt_nvbcq_enqueue(nxt_nvbcq_t volatile *q, nxt_nvbcq_atomic_t val)
--
src/nxt_nvbcq.h:113:static inline nxt_nvbcq_atomic_t
src/nxt_nvbcq.h-114-nxt_nvbcq_dequeue(nxt_nvbcq_t volatile *q)
--
src/nxt_unit.c:4436:static inline void
src/nxt_unit.c-4437-nxt_unit_process_lhq_pid(nxt_lvlhsh_query_t *lhq, pid_t *pid)
--
src/nxt_unit.c:6390:static inline void
src/nxt_unit.c-6391-nxt_unit_port_hash_lhq(nxt_lvlhsh_query_t *lhq,
--
src/nxt_unit_sptr.h:24:static inline void
src/nxt_unit_sptr.h-25-nxt_unit_sptr_set(nxt_unit_sptr_t *sptr, void *ptr)
--
src/nxt_unit_sptr.h:31:static inline void *
src/nxt_unit_sptr.h-32-nxt_unit_sptr_get(nxt_unit_sptr_t *sptr)
```

alejandro-colomar replied:
Fixed.

alejandro-colomar commented on src/nxt_clone.c:
```diff
@@ -426,14 +437,27 @@ nxt_clone_vldt_credential_gidmap(nxt_task_t *task,
+
+static inline
+id_t
+nxt_id_map_host2container(id_t host_id, const nxt_clone_map_entry_t *map)
```
Fixed.  Thanks!

hongzhidao commented:
Hi,
Is there any test case to show how the fixing work?

alejandro-colomar commented on src/nxt_clone.c:
```diff
@@ -395,12 +395,6 @@ nxt_clone_vldt_credential_gidmap(nxt_task_t *task,
         for (j = 0; j < map->size; j++) {
             m = map->map[j];
 
-            if (!base_ok && creds->base_gid >= (nxt_gid_t) m.container
-                && creds->base_gid < (nxt_gid_t) (m.container+m.size))
-            {
-                base_ok = 1;
-            }
```
Yes, this is part of the first patch, titled [Isolation: removed duplicate code.](https://github.com/nginx/unit/pull/800/commits/59150fa2f662d25a244f0e54a5fde9eaccf9db35) :)

It closes the following bug: <https://github.com/nginx/unit/issues/788>

alejandro-colomar commented:
> Hi, Is there any corresponding test case?

No.  I don't know how to write a pytest for it.  I'll share my manual tests tomorrow, and ask @andrey-zelenkov to transform it into a pytest.

alejandro-colomar commented:
Here go my manual tests that confirm that the bugfix actually fixes the bug:

```sh
alx@asus5775:~/etc/unitd$ id
uid=1000(alx) gid=1000(alx) groups=1000(alx),24(cdrom),25(floppy),29(audio),30(dip),44(video),46(plugdev),108(netdev),115(lpadmin),118(scanner)
alx@asus5775:~/etc/unitd$ echo {} | sudo unit ctl http PUT /config 2>/dev/null
{
	"success": "Reconfiguration done."
}
alx@asus5775:~/etc/unitd$ sudo unit ctl http GET /config 2>/dev/null </dev/null
{}
alx@asus5775:~/etc/unitd$ sudo unit ctl http PUT /config 2>/dev/null <idmap.json 
{
	"success": "Reconfiguration done."
}
alx@asus5775:~/etc/unitd$ sudo unit ctl http GET /config/applications/hello-php/isolation/ 2>/dev/null </dev/null
{
	"namespaces": {
		"cgroup": true,
		"credential": true,
		"mount": true,
		"network": true,
		"pid": true,
		"uname": true
	},

	"uidmap": [
		{
			"host": 1000,
			"container": 8001,
			"size": 1
		}
	],

	"gidmap": [
		{
			"host": 24,
			"container": 5024,
			"size": 977
		}
	]
}
alx@asus5775:~/etc/unitd$ <uidmap.json sed '/host/s/1000/1001/' | tee /dev/tty | sudo unit ctl http PUT /config/applications/hello-php/isolation/uidmap 2>/dev/null
[
    {
	"host": 1001,
	"container": 8001,
	"size": 1
    }
]
{
	"error": "Failed to apply new configuration."
}
alx@asus5775:~/etc/unitd$ <uidmap.json sed '/host/s/1000/999/' | tee /dev/tty | sudo unit ctl http PUT /config/applications/hello-php/isolation/uidmap 2>/dev/null
[
    {
	"host": 999,
	"container": 8001,
	"size": 1
    }
]
{
	"error": "Failed to apply new configuration."
}
alx@asus5775:~/etc/unitd$ <uidmap.json sed -e '/host/s/1000/999/' -e '/size/s/1/2/' | tee /dev/tty | sudo unit ctl http PUT /config/applications/hello-php/isolation/uidmap 2>/dev/null
[
    {
	"host": 999,
	"container": 8001,
	"size": 2
    }
]
{
	"success": "Reconfiguration done."
}
alx@asus5775:~/etc/unitd$ <gidmap.json sed '/host/s/24/25/' | tee /dev/tty | sudo unit ctl http PUT /config/applications/hello-php/isolation/gidmap 2>/dev/null
[
    {
	"host": 25,
	"container": 5024,
	"size": 977
    }
]
{
	"error": "Failed to apply new configuration."
}
alx@asus5775:~/etc/unitd$ <gidmap.json sed '/host/s/24/23/' | tee /dev/tty | sudo unit ctl http PUT /config/applications/hello-php/isolation/gidmap 2>/dev/null
[
    {
	"host": 23,
	"container": 5024,
	"size": 977
    }
]
{
	"error": "Failed to apply new configuration."
}
alx@asus5775:~/etc/unitd$ <gidmap.json sed -e '/host/s/24/23/' -e '/size/s/977/978/' | tee /dev/tty | sudo unit ctl http PUT /config/applications/hello-php/isolation/gidmap 2>/dev/null
[
    {
	"host": 23,
	"container": 5024,
	"size": 978
    }
]
{
	"success": "Reconfiguration done."
}
```

Oh boy, this `unit` script kept my sanity, did it?

hongzhidao commented:
Could you show a mini configuration that has an issue?
That we can apply it and it will reconfigure failed, then it works after merging your patch.

hongzhidao commented:
Could you show a mini configuration that has an issue?
That we can apply it and it will reconfigure failed, then it works after merging your patch.

alejandro-colomar commented:
> Could you show a mini configuration that has an issue? That we can apply it and it will reconfigure failed, then it works after merging your patch.

The following should work on every system.  Replace the "user" and "group" by your username, and a few other fields, for the application.

```json
{
    "listeners": {
        "*:8080": {
            "pass": "applications/hello-php"
        }
    },
    "applications": {
        "hello-php": {
            "type": "php",
            "group": "alx",
            "user": "alx",
            "root": "/home/alx/srv/www/",
            "isolation": {
                "namespaces": {
                    "cgroup": true,
                    "credential": true,
                    "mount": true,
                    "network": true,
                    "pid": true,
                    "uname": true
                },

                "uidmap": [
                    {
                        "host": 1000,
                        "container": 9876,
                        "size": 1
                    }
                ],

                "gidmap": [
                    {
                        "host": 1,
                        "container": 6789,
                        "size": 1000
                    }
                ]
            }

        }
    }
}
```

alejandro-colomar commented:
Changes:

-  Rebase to master (I forgot about interdiff issues :/).
-  Fix the tests.

alejandro-colomar commented:
Changes:

- Fix also unprivileged unit.  (It seems I had only fixed privileged unit)
- Add debug message.

I'll show manual tests for unprivileged unit.

alejandro-colomar commented:
Tests for unprivileged unit:

```json
alx@asus5775:~/etc/unitd$ cat idmap_unpriv.json 
{
    "access_log": "/tmp/access.log",
    "listeners": {
        "*:8080": {
            "pass": "applications/hello-php"
        }
    },
    "applications": {
        "hello-php": {
            "type": "php",
            "group": "man",
            "user": "man",
            "root": "/home/alx/srv/www/",
            "isolation": {
                "namespaces": {
                    "cgroup": true,
                    "credential": true,
                    "mount": true,
                    "network": true,
                    "pid": true,
                    "uname": true
                },

                "uidmap": [
                    {
                        "host": 6,
                        "container": 1234,
                        "size": 1
                    }
                ],

                "gidmap": [
                    {
                        "host": 12,
                        "container": 3456,
                        "size": 1
                    }
                ]
            }

        }
    }
} 
```
```sh
alx@asus5775:~/etc/unitd$ id man
uid=6(man) gid=12(man) groups=12(man)
```
```sh
alx@asus5775:~/etc/unitd$ ls -l /tmp/unit/
total 4
drwxr-xr-x 3 man man 4096 Jan  3 12:53 state
```
```sh
alx@asus5775:~/etc/unitd$ sudo -u man /opt/local/unit/sbin/unitd --control unix:/tmp/unit/control.unit.sock --state /tmp/unit/state --pid /tmp/unit/unit.pid --log /tmp/unit/unit.log 2>/dev/null; echo $?
0
```
```json
alx@asus5775:~/etc/unitd$ echo {} | sudo unit ctl http PUT /config 2>/dev/null
{
	"success": "Reconfiguration done."
}
```
```json
alx@asus5775:~/etc/unitd$ sudo unit ctl http GET /config 2>/dev/null </dev/null
{}
```
```json
alx@asus5775:~/etc/unitd$ sudo unit ctl http PUT /config 2>/dev/null <idmap_unpriv.json 
{
	"success": "Reconfiguration done."
}
```
```json
alx@asus5775:~/etc/unitd$ sed '/host/s/6/5/' <uidmap.json | tee /dev/tty | sudo unit ctl http PUT /config/applications/hello-php/isolation/uidmap 2>/dev/null
[
    {
	"host": 5,
	"container": 1234,
	"size": 1
    }
]
{
	"error": "Failed to apply new configuration."
}
```
```json
alx@asus5775:~/etc/unitd$ sed '/host/s/6/7/' <uidmap.json | tee /dev/tty | sudo unit ctl http PUT /config/applications/hello-php/isolation/uidmap 2>/dev/null
[
    {
	"host": 7,
	"container": 1234,
	"size": 1
    }
]
{
	"error": "Failed to apply new configuration."
}
```
```json
alx@asus5775:~/etc/unitd$ sed '/size/s/1/2/' <uidmap.json | tee /dev/tty | sudo unit ctl http PUT /config/applications/hello-php/isolation/uidmap 2>/dev/null
[
    {
	"host": 6,
	"container": 1234,
	"size": 2
    }
]
{
	"error": "Failed to apply new configuration."
}
```
```json
alx@asus5775:~/etc/unitd$ cat <uidmap.json | tee /dev/tty | sudo unit ctl http PUT /config/applications/hello-php/isolation/uidmap 2>/dev/null
[
    {
	"host": 6,
	"container": 1234,
	"size": 1
    }
]
{
	"success": "Reconfiguration done."
}
```
```json
alx@asus5775:~/etc/unitd$ sed '/host/s/12/11/' <gidmap.json | tee /dev/tty | sudo unit ctl http PUT /config/applications/hello-php/isolation/gidmap 2>/dev/null
[
    {
	"host": 11,
	"container": 2345,
	"size": 1
    }
]
{
	"error": "Failed to apply new configuration."
}
```
```json
alx@asus5775:~/etc/unitd$ sed '/host/s/12/13/' <gidmap.json | tee /dev/tty | sudo unit ctl http PUT /config/applications/hello-php/isolation/gidmap 2>/dev/null
[
    {
	"host": 13,
	"container": 2345,
	"size": 1
    }
]
{
	"error": "Failed to apply new configuration."
}
```
```json
alx@asus5775:~/etc/unitd$ sed '/size/s/1/2/' <gidmap.json | tee /dev/tty | sudo unit ctl http PUT /config/applications/hello-php/isolation/gidmap 2>/dev/null
[
    {
	"host": 12,
	"container": 2345,
	"size": 2
    }
]
{
	"error": "Failed to apply new configuration."
}
```
```json
alx@asus5775:~/etc/unitd$ cat <gidmap.json | tee /dev/tty | sudo unit ctl http PUT /config/applications/hello-php/isolation/gidmap 2>/dev/null
[
    {
	"host": 12,
	"container": 2345,
	"size": 1
    }
]
{
	"success": "Reconfiguration done."
}
```

alejandro-colomar commented:
Added a few commits:

-  Removed unnecessary restriction about unmapped supplementary groups in unprivileged unit.  Closes <https://github.com/nginx/unit/issues/820>.
-  Removed redundant checks for unprivileged unit.  The kernel will already fail, and the logs didn't improve significantly with our checks, so less code is better code.  In fact, I found that the restriction above was unnecessary when simplifing the code.

alejandro-colomar commented:
Changes:

-  Fixed a few minor things for unprivileged unitd.


I can confirm that the tests I performed before still work (after these fixes).


These patches might break some things with `rootfs` that @i4ki mentioned yesterday.  I haven't been able to test that yet.  I want to make sure I don't break that, so hold on a little bit for more updates.

alejandro-colomar commented:
I can't manage to get `rootfs` working with the tests I'm using.  Not even completely removing id maps from the configuration.  I guess we can merge these patches, and maybe in the future we can revisit it and improve it.

alejandro-colomar commented:
Changes:

-  Rebase to master

```
$ git range-diff 970a9812^..4e5f17c3 master..HEAD
 1:  970a9812 !  1:  6c90f149 Isolation: removed duplicate code.
    @@ Commit message
         the result will always be the same.
     
         Closes: <https://github.com/nginx/unit/issues/799>
    +    Cc: Tiago Natel de Moura <t.nateldemoura@f5.com>
         Signed-off-by: Alejandro Colomar <alx@nginx.com>
     
      ## src/nxt_clone.c ##
 2:  1484b76a !  2:  b5ad73ad Isolation: moved code to have logs next to their test.
    @@ Commit message
         and then reporting if any of them failed, prefer testing one, reporting,
         testing the other, and reporting.
     
    +    Cc: Tiago Natel de Moura <t.nateldemoura@f5.com>
         Signed-off-by: Alejandro Colomar <alx@nginx.com>
     
      ## src/nxt_clone.c ##
 3:  c97d95cc !  3:  850ab43d Isolation: fixed mapping of user and group IDs.
    @@ Commit message
     
         Closes: <https://github.com/nginx/unit/issues/788>
         Reported-by: <bsidhu009@gmail.com>
    +    Cc: Tiago Natel de Moura <t.nateldemoura@f5.com>
         Signed-off-by: Alejandro Colomar <alx@nginx.com>
     
      ## docs/changes.xml ##
    -@@ docs/changes.xml: NGINX Unit updated to 1.30.0.
    -          date="" time=""
    -          packager="Nginx Packaging &lt;nginx-packaging@f5.com&gt;">
    +@@ docs/changes.xml: PHP error handling (added missing 403 and 404 errors).
    + </para>
    + </change>
      
     +<change type="bugfix">
     +<para>
    @@ docs/changes.xml: NGINX Unit updated to 1.30.0.
     +
      </changes>
      
    - <changes apply="unit-php
    + 
     
      ## src/nxt_clone.c ##
     @@
    - #include <nxt_conf.h>
    - #include <nxt_clone.h>
    - 
    -+
    - #if (NXT_HAVE_CLONE)
    - 
    - pid_t
    -@@ src/nxt_clone.c: nxt_clone(nxt_int_t flags)
      
      #if (NXT_HAVE_CLONE_NEWUSER)
      
 4:  570b95f4 !  4:  8786765b Isolation: factored out complex and repetitive check.
    @@ Metadata
      ## Commit message ##
         Isolation: factored out complex and repetitive check.
     
    +    Cc: Tiago Natel de Moura <t.nateldemoura@f5.com>
         Signed-off-by: Alejandro Colomar <alx@nginx.com>
     
      ## src/nxt_clone.c ##
    -@@ src/nxt_clone.c: nxt_clone(nxt_int_t flags)
    +@@
      
      static inline id_t nxt_id_map_host2container(id_t host_id,
          const nxt_clone_map_entry_t *map);
 5:  e6679d7a !  5:  603c3548 Isolation: minor optimization.
    @@ Commit message
         some fields.  It's faster (and conceptually simpler) to just take a
         pointer.  Also, make it const, since we're not modifying the contents.
     
    +    Cc: Tiago Natel de Moura <t.nateldemoura@f5.com>
         Signed-off-by: Alejandro Colomar <alx@nginx.com>
     
      ## src/nxt_clone.c ##
 6:  62128bac !  6:  c7395cc7 Tests: running tests only one.
    @@ Metadata
     Author: Alejandro Colomar <alx@nginx.com>
     
      ## Commit message ##
    -    Tests: running tests only one.
    +    Tests: running tests only once.
     
         We were running tests twice in some cases.  This seems a glitch, and got
         me for some time wondering why some apparently-fine tests were failing.
         It was that privileged tests were running twice, and since it's a
         destructive test, the second run was being run with invalid data.
     
    +    Cc: Tiago Natel de Moura <t.nateldemoura@f5.com>
         Signed-off-by: Alejandro Colomar <alx@nginx.com>
     
      ## src/test/nxt_clone_test.c ##
 7:  7f26d8ca !  7:  f5320542 Tests: fixed tests for uidmap and gidmap.
    @@ Commit message
     
         The tests were mostly wrong.  This is almost a rewrite.
     
    +    Cc: Tiago Natel de Moura <t.nateldemoura@f5.com>
         Signed-off-by: Alejandro Colomar <alx@nginx.com>
     
      ## src/test/nxt_clone_test.c ##
 8:  50a962c8 !  8:  5adb14e3 Isolation: added debug message for id mappings.
    @@ Commit message
     
         Show which concrete mappings are performed.
     
    +    Cc: Tiago Natel de Moura <t.nateldemoura@f5.com>
         Signed-off-by: Alejandro Colomar <alx@nginx.com>
     
      ## src/nxt_clone.c ##
    -@@ src/nxt_clone.c: nxt_clone(nxt_int_t flags)
    +@@
      
      #if (NXT_HAVE_CLONE_NEWUSER)
      
 9:  fe0191cb !  9:  9b89e709 Isolation: fixed uidmap for unprivileged unit.
    @@ Commit message
         Unprivileged unit was checking the mapping was correct, but then wasn't
         applying it.
     
    +    Cc: Tiago Natel de Moura <t.nateldemoura@f5.com>
         Signed-off-by: Alejandro Colomar <alx@nginx.com>
     
      ## src/nxt_clone.c ##
10:  f076d8c0 ! 10:  465825cc Isolation: using generic code path.
    @@ Commit message
         code for unprivileged unit, and let it fall through to the same code
         path that handles privileged unit.
     
    +    Cc: Tiago Natel de Moura <t.nateldemoura@f5.com>
         Signed-off-by: Alejandro Colomar <alx@nginx.com>
     
      ## src/nxt_clone.c ##
11:  974613ae ! 11:  c5cfe6f0 Isolation: allowing groups to remain unmapped in unprivileged unit.
    @@ Commit message
         the kernel is not backing us, but when unprivileged, we can allow it.
     
         Closes: <https://github.com/nginx/unit/issues/820>
    +    Cc: Tiago Natel de Moura <t.nateldemoura@f5.com>
         Signed-off-by: Alejandro Colomar <alx@nginx.com>
     
      ## src/nxt_clone.c ##
12:  df8b06d5 ! 12:  112599f8 Isolation: removed redundant checks for unprivileged unitd.
    @@ Commit message
         Since we don't check in our code, but rather rely on the kernel doing
         the checks, some C tests don't make sense anymore.
     
    +    Cc: Tiago Natel de Moura <t.nateldemoura@f5.com>
         Signed-off-by: Alejandro Colomar <alx@nginx.com>
     
      ## src/nxt_clone.c ##
13:  4e5f17c3 ! 13:  f756b085 Isolation: reversed logic to improve readability.
    @@ Metadata
      ## Commit message ##
         Isolation: reversed logic to improve readability.
     
    +    Cc: Tiago Natel de Moura <t.nateldemoura@f5.com>
         Signed-off-by: Alejandro Colomar <alx@nginx.com>
     
      ## src/nxt_clone.c ##

```
